### ISSUE TITLE
Dependencies are rebuild upon changes to rust source

### ISSUE THREAD

**Issue by Panke** [ipetkov/crane#194]
I have a workspace cargo project (i.e. the toplevel Cargo.toml only contains a `[workspace]` section
and the following flake.nix:

```nix
{
  description = "Test";
  inputs = {
    nixpkgs.url = github:nixos/nixpkgs/nixpkgs-unstable;
    flake-compat = {
      url = "github:edolstra/flake-compat";
      flake = false;
    };
    crane = {
      url = "github:ipetkov/crane";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };
  outputs = { self, nixpkgs, crane, ... }:
    let
      system = "x86_64-linux";
      pkgs = nixpkgs.legacyPackages.${system};
      craneLib = crane.lib.${system};
      buildInputs = [
            pkgs.openssl pkgs.pkg-config pkgs.nodePackages.tailwindcss pkgs.cargo pkgs.rustc pkgs.cargo-insta
      ];
      my-crate = craneLib.buildPackage {
        src = ./.;
        buildInputs = buildInputs;
      };
    in
    {
      packages.${system}.default = my-crate;
      devShell.${system} =
        pkgs.mkShell {
          buildInputs = [ pkgs.nixpkgs-fmt pkgs.rustfmt ] ++ buildInputs;
        };
    };
}
```

Whenever I change one of the contained rust files all dependencies are rebuild. I already found issue #25 , but the advice in there didn't help me (besides realizing that I do not need to split the dependencies apart). 

I am a bit at a loss on how to debug this further.

**Comment by ipetkov** [comment]
Hi @Panke thanks for the report!

We go to some lengths to ensure that the only inputs which can cause a rebuild of the dependencies are changes to `Cargo.toml`, `Cargo.lock`, or `.config/cargo.toml`. I can't confirm without looking at the actual flake of the project but I noticed the flake lists `src = ./.;` which will cause a rebuild (of the final crate) when _any_ sources change, including changes to `flake.nix`.

My recommendation is to try using `src = craneLib.cleanCargoSource ./.;` and see if that helps!

**Comment by dennybritz** [comment]
I'm seeing the same issue. A change to *any* file, even if it's ignored in the `src` triggers a rebuild of the dependencies. I don't remember this happening previously so I tracked it down to this commit: https://github.com/ipetkov/crane/commit/70fcce8db07c5121b02c2f1484768a3cff59edac

If I pin the `crane` revision to a previous commit this doesn't happen, so I think there may be a bug in the changes above?

**Comment by ipetkov** [comment]
@dennybritz ah thank you for narrowing it down! Yes that commit introduced the regression, we need to throw the string context away. Should have a fix up in a bit

**Comment by Panke** [comment]
Thanks everyone! I updated to the newest version of crane, but it still needlessly rebuild the dependency package for me. I want to debug this myself, but I am quite new to nix: any hints on how to figure out why a derivation thinks it is out of date?

**Comment by Panke** [comment]
I made a reduced test case that reproduces the problem. Looking through the involved *.drv files of the involved derivations leads to a src of the deps package containing rust files. I'll upload the reduced test cases when I get stuck.

**Comment by ipetkov** [comment]
> Thanks everyone! I updated to the newest version of crane, but it still needlessly rebuild the dependency package for me.

@Panke did the dependencies get rebuilt _just once_ or are they getting rebuilt after any change?

It is expected to potentially rebuild all dependencies if any of the build scripts have changed since the last time the flake was updated (Nix considers them inputs to the derivation, so changing any of the build scripts will result in rebuilding the dependency derivation too). If further changes to the source files _are still rebuilding everything from scratch_ then we might have another regression, so I wanted to double check there

**Comment by Panke** [comment]
Confirm fix with #242
